glslc/src/main.cc

```diff
@@ -165,35 +177,54 @@ int main(int argc, char** argv) {
         }
       }
     } else if (arg == "-c") {
-      compiler.SetIndividualCompilationMode();
+      compiler.SetIndividualCompilationFlag();
     } else if (arg == "-E") {
-      compiler.SetPreprocessingOnlyMode();
+      compiler.options().SetPreprocessingOnlyMode();
+      compiler.SetPreprocessingOnlyFlag();
     } else if (arg == "-S") {
-      compiler.SetDisassemblyMode();
+      compiler.options().SetDisassemblyMode();
+      compiler.SetDisassemblyFlag();
     } else if (arg.starts_with("-D")) {
       const size_t length = arg.size();
       if (length <= 2) {
         std::cerr << "glslc: error: argument to '-D' is missing" << std::endl;
       } else {
         const string_piece argument = arg.substr(2);
-        size_t name_length = argument.find_first_of('=');
-        const string_piece name = argument.substr(0, name_length);
-        if (name.starts_with("GL_")) {
+        // Get the exact length of the macro string.
+        size_t equal_sign_loc = argument.find_first_of('=');
+        size_t name_length = equal_sign_loc != shaderc_util::string_piece::npos
+                                 ? equal_sign_loc
+                                 : (arg.size() - strlen("-D"));
+        const string_piece name_piece = argument.substr(0, name_length);
+        size_t argument_length = argument.size() - (argument.back() == '=' ? 1 : 0);
+        if (name_piece.starts_with("GL_")) {
           std::cerr
               << "glslc: error: names beginning with 'GL_' cannot be defined: "
               << arg << std::endl;
           return 1;
         }
-        if (name.find("__") != string_piece::npos) {
+        if (name_piece.find("__") != string_piece::npos) {
           std::cerr
               << "glslc: warning: names containing consecutive underscores "
                  "are reserved: "
               << arg << std::endl;
         }
+
+        // We have to insert a null terminator at the end of name, as we are
+        // using libshaderc's API instead of libshaderc_util's. Libshaderc API
+        // assumes a const char* string, while libshaderc_util assume a
+        // string_piece one. The original implementation doesn't append '\0' but
+        // works fine with libshaderc_util because it has 'end_' property. But
+        // libshaderc api doesn't know that so the macro 'name' will include the
+        // 'value' if we don't insert '\0'.
+
+        macro_names.emplace_back(name_piece.data(), name_length);
+        macro_values.emplace_back(name_length < argument_length
```

Comment:
Use this condition instead: `equal_sign_loc == npos || equal_sign_loc == argument.size() - 1`.